utils/pattern/src/double.rs

```diff
@@ -0,0 +1,526 @@
+/// let store = Pattern::<DoublePlaceholder, _>::try_from_str("Hello, {0} and {1}!")
+///     .unwrap()
+///     .take_store();
+///
+/// assert_eq!("\x10\x1BHello,  and !", store);
+/// ```
+///
+/// Explanation of the lead code points:
+///
+/// 1. `\x10` is placeholder 0 at index 7: ((7 + 1) << 1) | 0
+/// 2. `\x1B` is placeholder 1 at index 12: ((12 + 1) << 1) | 1
+///
+/// Example patterns supported by this backend:
+///
+/// ```
+/// use icu_pattern::Pattern;
+/// use icu_pattern::DoublePlaceholder;
+///
+/// // Single numeric placeholder (note, "5" is used):
+/// assert_eq!(
+///     Pattern::<DoublePlaceholder, _>::try_from_str("{0} days ago")
+///         .unwrap()
+///         .interpolate_to_string([5, 7]),
+///     "5 days ago",
+/// );
+///
+/// // No placeholder (note, the placeholder value is never accessed):
+/// assert_eq!(
+///     Pattern::<DoublePlaceholder, _>::try_from_str("yesterday")
+///         .unwrap()
+///         .interpolate_to_string(["foo", "bar"]),
+///     "yesterday",
+/// );
+///
+/// // Escaped placeholder and a placeholder value 1 (note, "bar" is used):
+/// assert_eq!(
+///     Pattern::<DoublePlaceholder, _>::try_from_str("'{0}' {1}")
+///         .unwrap()
+///         .interpolate_to_string(("foo", "bar")),
+///     "{0} bar",
+/// );
+///
+/// // Pattern with the placeholders in the opposite order:
+/// assert_eq!(
+///     Pattern::<DoublePlaceholder, _>::try_from_str("A {1} B {0} C")
+///         .unwrap()
+///         .interpolate_to_string(("D", "E")),
+///     "A E B D C",
+/// );
+///
+/// // No literals, only placeholders:
+/// assert_eq!(
+///     Pattern::<DoublePlaceholder, _>::try_from_str("{1}{0}")
+///         .unwrap()
+///         .interpolate_to_string((1, 2)),
+///     "21",
+/// );
+/// ```
+///
+/// [`Pattern::interpolate()`]: crate::Pattern::interpolate
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[allow(clippy::exhaustive_enums)] // Empty Enum
+pub enum DoublePlaceholder {}
+
+impl crate::private::Sealed for DoublePlaceholder {}
+
+impl PatternBackend for DoublePlaceholder {
+    type PlaceholderKey = DoublePlaceholderKey;
+    type Store = str;
+    type Iter<'a> = DoublePlaceholderPatternIterator<'a>;
+
+    fn validate_store(store: &Self::Store) -> Result<(), Error> {
+        let mut chars = store.chars();
+        let ph_first_char = chars.next().ok_or(Error::InvalidPattern)?;
+        let ph_second_char = chars.next().ok_or(Error::InvalidPattern)?;
+        let initial_offset = ph_first_char.len_utf8() + ph_second_char.len_utf8();
+        let ph_first = DoublePlaceholderInfo::from_char(ph_first_char);
+        let ph_second = DoublePlaceholderInfo::from_char(ph_second_char);
+        if ph_first.key == ph_second.key {
+            debug_assert!(false, "ph_first.key == ph_second.key");
+            return Err(Error::InvalidPattern);
+        }
+        if ph_first.offset > ph_second.offset && ph_second.offset > 0 {
```

Comment:
issue: be strict in what you produce but lenient in what you accept

Reply:
Yes, `validate_store` is useful, but the invariant that the two placeholders are in order is not required, is it? We can accept data that is out of order.

Reply:
It's enforcing that the offsets are in order, not that the placeholders are in order.